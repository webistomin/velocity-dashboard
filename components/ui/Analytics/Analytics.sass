+b('analytics')
  +e('grid')
    display: grid
    justify-content: start
    grid-template-columns: repeat(12, 1fr)
    grid-gap: rem(14px)

    +media($screen-sm)
      grid-gap: rem(24px)

    +media($screen-lg)
      grid-gap: rem(34px)

  +e('block')
    &:first-of-type
      grid-column: 12 span
      grid-row: 4 span

    &:nth-of-type(2),
    &:nth-of-type(3),
    &:nth-of-type(4),
    &:nth-of-type(5)
      grid-column: 12 span
      grid-row: 1 span

    &:nth-of-type(6)
      grid-column: 12 span
      grid-row: 2 span

    &:nth-of-type(7)
      grid-column: 12 span
      grid-row: 3 span

    +media($screen-sm)
      &:nth-of-type(2),
      &:nth-of-type(3),
      &:nth-of-type(4),
      &:nth-of-type(5)
        grid-column: 6 span
        grid-row: 1 span

    +media($screen-lg)
      &:nth-of-type(2),
      &:nth-of-type(3),
      &:nth-of-type(4),
      &:nth-of-type(5)
        grid-column: 3 span
        grid-row: 1 span

      &:nth-of-type(6)
        grid-column: 4 span
        grid-row: 3 span

      &:nth-of-type(7)
        grid-column: 8 span
        grid-row: 3 span


