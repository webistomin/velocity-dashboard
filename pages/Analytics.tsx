import { VueComponent } from 'types/vue-components';
import { Component } from 'nuxt-property-decorator';
import { VNode } from 'vue';
import Analytics from 'components/ui/Analytics';
import { Context } from '@nuxt/types';
import { serverUrls } from 'common/urls/serverUrls';
import { IPageAnalytics, IPageAnalyticsResponse } from 'common/types/pages/analytics';

@Component({
  name: 'AnalyticsPage',
  middleware: ['auth-guard', 'auth-admin'],
})
export default class AnalyticsPage extends VueComponent {
  public content: IPageAnalytics | null = null;

  public head() {
    return {
      title: 'Velocity:: Analytics',
    };
  }

  public async asyncData(context: Context) {
    try {
      const data: IPageAnalyticsResponse = await context.$axios.$get(serverUrls.pages.analytics);
      return { content: data.content };
    } catch (error) {
      console.log(error);
    }
  }

  render(): VNode {
    return (
      <main class='page-content page-content_offset'>
        <Analytics content={this.content} />
      </main>
    );
  }
}
