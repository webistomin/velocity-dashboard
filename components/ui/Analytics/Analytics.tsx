import { VueComponent } from 'types/vue-components';
import { Component, Prop } from 'nuxt-property-decorator';
import { VNode } from 'vue';
import { ChartData, ChartOptions } from 'chart.js';
import formatDistance from 'date-fns/formatDistance';
// @ts-ignore
import LazyHydrate from 'vue-lazy-hydration';

import BaseBlock from 'components/base/BaseBlock';
import BaseLineChart from 'components/base/BaseLineChart';
import BaseStat from 'components/base/BaseStat';
import BaseList from 'components/base/BaseList';
import BaseBarGraph from 'components/base/BaseBarCraph';
import { IPageAnalytics } from 'common/types/pages/analytics';
import { IBaseList } from 'components/base/BaseList/BaseList';
import BaseEmpty from 'components/base/BaseEmpty';

import './Analytics.sass';

export interface IAnalyticsPageProps {
  content: IPageAnalytics | null;
}

@Component({
  name: 'Analytics',
  components: { LazyHydrate },
})
export default class Analytics extends VueComponent<IAnalyticsPageProps> {
  @Prop({ required: true })
  private readonly content!: IAnalyticsPageProps['content'];

  public lineOptions: ChartOptions = {
    responsive: true,
    maintainAspectRatio: false,
    scales: {
      yAxes: [
        {
          gridLines: {
            color: '#B0BAC9',
            borderDash: [3, 3],
          },
          ticks: {
            suggestedMin: 0,
            suggestedMax: 125,
            fontColor: '#B0BAC9',
            callback: (value) => {
              let tempVal = value.toString();
              if (tempVal.length >= 4) tempVal = tempVal.substring(0, tempVal.length - 3) + 'k ';
              return '$' + tempVal;
            },
          },
        },
      ],
      xAxes: [
        {
          gridLines: {
            color: '#B0BAC9',
            borderDash: [3, 3],
          },
          ticks: {
            fontColor: '#B0BAC9',
          },
        },
      ],
    },
    legend: {
      display: true,
      align: 'end',
      labels: {
        boxWidth: 10,
        usePointStyle: true,
        padding: 20,
        fontColor: '#B0BAC9',
      },
    },
    elements: {
      point: {
        radius: 5,
        hoverRadius: 4,
        borderWidth: 3,
        hoverBorderWidth: 3,
      },
    },
  };

  public mixedOptions: ChartOptions = {
    responsive: true,
    maintainAspectRatio: false,
    scales: {
      yAxes: [
        {
          gridLines: {
            color: '#B0BAC9',
            borderDash: [3, 3],
          },
          ticks: {
            suggestedMin: 0,
            suggestedMax: 125,
            fontColor: '#B0BAC9',
          },
        },
      ],
      xAxes: [
        {
          gridLines: {
            color: '#B0BAC9',
            borderDash: [3, 3],
          },
          ticks: {
            fontColor: '#B0BAC9',
          },
        },
      ],
    },
    legend: {
      display: true,
      align: 'end',
      labels: {
        boxWidth: 10,
        usePointStyle: true,
        padding: 20,
        fontColor: '#B0BAC9',
      },
    },
    elements: {
      point: {
        radius: 5,
        hoverRadius: 4,
        borderWidth: 3,
        hoverBorderWidth: 3,
        backgroundColor: '#FFF',
      },
    },
  };

  public get getRevenueChartData(): ChartData | undefined {
    const revenue = this.content?.revenue;

    if (!revenue) {
      return undefined;
    }

    return {
      labels: revenue.labels,
      datasets: [
        {
          label: 'Revenue',
          borderColor: '#2E5BFF',
          pointBackgroundColor: '#FFF',
          fill: true,
          lineTension: 0,
          data: revenue.data[0],
        },
        {
          label: 'Trips',
          borderColor: '#8C54FF',
          pointBackgroundColor: '#FFF',
          fill: true,
          lineTension: 0,
          data: revenue.data[1],
        },
      ],
    };
  }

  public get getLatestTrips(): IBaseList[] | undefined {
    return this.content?.latestTrips?.trips?.map((trip) => {
      return {
        image: '/img/avatar-placeholder.svg',
        name: `${trip.trip.startAddress} ${trip.trip.endAddress}`,
        text: `${trip.trip.endTime ? formatDistance(new Date(trip.trip.endTime), new Date()) : ''}`,
        statistic: `${trip.trip.distance} miles`,
        price: `$${trip.trip.price}`,
      };
    });
  }

  public get getTripsByWeekdayChartData(): ChartData | undefined {
    const tripsByWeekday = this.content?.tripsByWeekday;

    if (!tripsByWeekday) {
      return undefined;
    }

    return {
      labels: ['Mon', 'Tue', 'Fri', 'Thu', 'Wed', 'Sat', 'Sun'],
      datasets: [
        {
          label: 'Average',
          borderColor: '#F7C137',
          fill: false,
          lineTension: 0,
          data: tripsByWeekday.datasets[0],
          type: 'line',
        },
        {
          barPercentage: 0.5,
          categoryPercentage: 0.3,
          label: 'Comfort',
          backgroundColor: '#2E5BFF',
          data: tripsByWeekday.datasets[1],
        },
        {
          barPercentage: 0.5,
          categoryPercentage: 0.3,
          label: 'Premium',
          backgroundColor: '#8C54FF',
          data: tripsByWeekday.datasets[2],
        },
      ],
    };
  }

  public render(): VNode {
    return (
      <section class='analytics'>
        <div class='container'>
          <div class='analytics__grid'>
            <BaseBlock class='analytics__block' title='Revenue'>
              {this.getRevenueChartData ? (
                <BaseLineChart
                  chartData={this.getRevenueChartData}
                  options={this.lineOptions}
                  gradients={[
                    {
                      red: 46,
                      green: 91,
                      blue: 255,
                    },
                    {
                      red: 140,
                      green: 84,
                      blue: 255,
                    },
                  ]}
                />
              ) : (
                <BaseEmpty />
              )}
            </BaseBlock>

            <BaseBlock class='analytics__block' title='Vehicles on track'>
              {this.content?.vehiclesOnTrack ? (
                <BaseStat
                  value={this.content.vehiclesOnTrack.todayCount}
                  prevValue={this.content.vehiclesOnTrack.yesterdayCount}
                  measure='$'
                  icon='icon-check'
                  color='green'
                  align='row'
                />
              ) : (
                <BaseEmpty />
              )}
            </BaseBlock>

            <BaseBlock class='analytics__block' title='Distance driven'>
              {this.content?.distanceDriven ? (
                <BaseStat
                  value={this.content.distanceDriven.todayCount}
                  prevValue={this.content.distanceDriven.yesterdayCount}
                  measure='mi'
                  icon='icon-pin'
                  color='blue'
                  align='row'
                />
              ) : (
                <BaseEmpty />
              )}
            </BaseBlock>

            <BaseBlock class='analytics__block' title='Energy consumed'>
              {this.content?.energyConsumed ? (
                <BaseStat
                  value={this.content.energyConsumed.todayCount}
                  prevValue={this.content.energyConsumed.yesterdayCount}
                  measure='kWh'
                  icon='icon-lightning'
                  color='purple'
                  align='row'
                />
              ) : (
                <BaseEmpty />
              )}
            </BaseBlock>

            <BaseBlock class='analytics__block' title='Total drive time'>
              {this.content?.totalDriveTime ? (
                <BaseStat
                  value={this.content.totalDriveTime.todayCount}
                  prevValue={this.content.totalDriveTime.yesterdayCount}
                  measure='h'
                  icon='icon-graph'
                  color='yellow'
                  align='row'
                />
              ) : (
                <BaseEmpty />
              )}
            </BaseBlock>

            <LazyHydrate when-visible>
              <BaseBlock class='analytics__block' title='Latest trips'>
                {this.getLatestTrips ? <BaseList list={this.getLatestTrips} /> : <BaseEmpty />}
              </BaseBlock>
            </LazyHydrate>

            <LazyHydrate when-visible>
              <BaseBlock class='analytics__block' title='Trips by weekday'>
                {this.getTripsByWeekdayChartData ? (
                  <BaseBarGraph chartData={this.getTripsByWeekdayChartData} options={this.mixedOptions} />
                ) : (
                  <BaseEmpty />
                )}
              </BaseBlock>
            </LazyHydrate>
          </div>
        </div>
      </section>
    );
  }
}
