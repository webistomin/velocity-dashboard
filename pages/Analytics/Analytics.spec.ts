import { shallowMount } from '@vue/test-utils';
import Analytics from './Analytics';

describe('Analytics page', () => {
  const factory = (propsData = {}, options = {}) => {
    return shallowMount(Analytics, {
      propsData: {
        ...propsData,
      },
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });
});
