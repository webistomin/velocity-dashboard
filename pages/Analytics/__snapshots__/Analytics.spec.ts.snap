// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics page Matches snapshot 1`] = `
"<main class=\\"page-content page-content_offset\\">
  <analytics-stub></analytics-stub>
</main>"
`;
